File: Kalorie/Kalorie/Features/Moderation/ModerationReviewViewModel.swift
//
//  ModerationReviewViewModel.swift
//  Kalorie
//
//  Created by Josef Antoni on 10.09.2026.
//

import Foundation
import UIKit

final class ModerationReviewViewModel: ObservableObject, NutritionLabelPrefilling {

    // MARK: - Properties

    @Published var formInput: FoodItemFormInput
    @Published private(set) var state: LoadingState<Void> = .idle
    @Published var alertItem: AlertItem?
    @Published var isRejectSheetVisible = false
    @Published var rejectReason = ""
    @Published private(set) var shouldDismiss = false
    @Published var recognizedFields: Set<FoodItemFormField> = []
    @Published var isRecognizingNutritionLabel = false
    @Published var isNutritionLabelCameraVisible = false
    @Published var nutritionLabelCameraHint: String?
    @Published private(set) var similarCatalogueItems: [FoodItemDomain] = []
    @Published private(set) var isSimilarCatalogueItemsSectionAvailable = true

    let submittedAt: Date
    let rejectReasonIfAny: String?

    var showsSimilarCatalogueItemsSection: Bool { submission.barcode == nil }

    private let submission: FoodItemSubmissionDomain
    private let approveSubmission: any ApproveSubmissionUseCaseProtocol
    private let rejectSubmission: any RejectSubmissionUseCaseProtocol
    private let searchFoodItems: any SearchFoodItemsUseCaseProtocol
    private let recognizeNutritionLabelUseCase: any RecognizeNutritionLabelUseCaseProtocol
    private let cameraAuthorizationProvider: any CameraAuthorizationProviderProtocol
    private let onResolved: () -> Void

    // MARK: - Init

    init(
        submission: FoodItemSubmissionDomain,
        approveSubmission: any ApproveSubmissionUseCaseProtocol,
        rejectSubmission: any RejectSubmissionUseCaseProtocol,
        searchFoodItems: any SearchFoodItemsUseCaseProtocol,
        recognizeNutritionLabel: any RecognizeNutritionLabelUseCaseProtocol,
        cameraAuthorizationProvider: any CameraAuthorizationProviderProtocol,
        onResolved: @escaping () -> Void
    ) {
        self.submission = submission
        self.formInput = FoodItemFormInput(item: submission.item)
        self.submittedAt = submission.submittedAt
        self.rejectReasonIfAny = submission.rejectReason
        self.approveSubmission = approveSubmission
        self.rejectSubmission = rejectSubmission
        self.searchFoodItems = searchFoodItems
        self.recognizeNutritionLabelUseCase = recognizeNutritionLabel
        self.cameraAuthorizationProvider = cameraAuthorizationProvider
        self.onResolved = onResolved
    }

    // MARK: - Functions

    @MainActor
    func onAppear() async {
        guard showsSimilarCatalogueItemsSection else { return }
        do {
            similarCatalogueItems = try await searchFoodItems(query: submission.item.czName)
        } catch {
            Log.warning(error, category: Constants.LogCategory.moderation)
            isSimilarCatalogueItemsSectionAvailable = false
        }
    }

    @MainActor
    func onNutritionLabelCaptured(_ image: UIImage, liveBarcode: String?) async {
        await recognizeNutritionLabel(from: image, liveBarcode: liveBarcode, using: recognizeNutritionLabelUseCase)
    }

    @MainActor
    func onNutritionLabelCameraTapped() async {
        await openNutritionLabelCamera(using: cameraAuthorizationProvider) {
            alertItem = AlertItem(title: L10n.AddFood.cameraPermissionAlert)
        }
    }

    @MainActor
    func onApproveTapped() async {
        state = .loading
        defer { state = .loaded }
        let editedItem = formInput.asFoodItemDomain(date: submission.item.date)
        do {
            try await approveSubmission(submission: submission, item: editedItem)
            onResolved()
            shouldDismiss = true
        } catch ApproveSubmissionError.alreadyResolved {
            onResolved()
            alertItem = AlertItem(title: L10n.Moderation.errorAlreadyResolved)
            shouldDismiss = true
        } catch ApproveSubmissionError.changedSinceReview {
            onResolved()
            alertItem = AlertItem(title: L10n.Moderation.errorChangedSinceReview)
            shouldDismiss = true
        } catch {
            Log.error(error, category: Constants.LogCategory.moderation)
            switch error as? CreateFoodItemError {
            case .invalidCode:
                alertItem = AlertItem(title: L10n.AddFood.errorInvalidCode)
            case .invalidName:
                alertItem = AlertItem(title: L10n.AddFood.errorInvalidName)
            case .invalidCalories:
                alertItem = AlertItem(title: L10n.AddFood.errorInvalidCalories)
            case .invalidWeight:
                alertItem = AlertItem(title: L10n.AddFood.errorInvalidWeight)
            case .invalidPortion(let portionError):
                alertItem = AlertItem(title: portionError.alertTitle)
            case .itemAlreadyExists:
                alertItem = AlertItem(title: L10n.Moderation.errorAlreadyExists)
            case nil:
                alertItem = AlertItem(title: L10n.Common.errorUnknown)
            }
        }
    }

    @MainActor
    func onRejectConfirmed() async {
        state = .loading
        defer { state = .loaded }
        do {
            try await rejectSubmission(submission, reason: rejectReason)
            onResolved()
            shouldDismiss = true
        } catch RejectSubmissionError.reasonRequired {
            alertItem = AlertItem(title: L10n.Moderation.errorReasonRequired)
        } catch RejectSubmissionError.alreadyResolved {
            onResolved()
            alertItem = AlertItem(title: L10n.Moderation.errorAlreadyResolved)
            shouldDismiss = true
        } catch RejectSubmissionError.changedSinceReview {
            onResolved()
            alertItem = AlertItem(title: L10n.Moderation.errorChangedSinceReview)
            shouldDismiss = true
        } catch {
            Log.error(error, category: Constants.LogCategory.moderation)
            alertItem = AlertItem(title: L10n.Common.errorUnknown)
        }
    }
}

File: Android/app/src/main/java/antoni/kalorie/features/moderation/ModerationReviewViewModel.kt
package antoni.kalorie.features.moderation

import androidx.lifecycle.ViewModel
import antoni.kalorie.R
import antoni.kalorie.components.FoodItemFormField
import antoni.kalorie.core.models.FoodItemDomain
import antoni.kalorie.core.models.FoodItemSubmissionDomain
import antoni.kalorie.core.nutritionlabelrecognition.NutritionLabelImage
import antoni.kalorie.core.nutritionlabelrecognition.RecognizeNutritionLabelUseCaseProtocol
import antoni.kalorie.core.usecases.ApproveSubmissionError
import antoni.kalorie.core.usecases.ApproveSubmissionUseCaseProtocol
import antoni.kalorie.core.usecases.CreateFoodItemError
import antoni.kalorie.core.usecases.RejectSubmissionError
import antoni.kalorie.core.usecases.RejectSubmissionUseCaseProtocol
import antoni.kalorie.core.usecases.SearchFoodItemsUseCaseProtocol
import antoni.kalorie.core.utils.AlertItem
import antoni.kalorie.core.utils.Constants
import antoni.kalorie.core.utils.LoadingState
import antoni.kalorie.core.utils.Log
import antoni.kalorie.core.utils.NutritionLabelPrefilling
import antoni.kalorie.core.utils.isLoading
import antoni.kalorie.features.addfoodsheet.FoodItemFormInput
import java.time.Instant
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow

class ModerationReviewViewModel(
    private val submission: FoodItemSubmissionDomain,
    private val approveSubmission: ApproveSubmissionUseCaseProtocol,
    private val rejectSubmission: RejectSubmissionUseCaseProtocol,
    private val searchFoodItems: SearchFoodItemsUseCaseProtocol,
    private val recognizeNutritionLabelUseCase: RecognizeNutritionLabelUseCaseProtocol,
    private val onResolved: () -> Unit,
) : ViewModel(), NutritionLabelPrefilling {

    // MARK: - Properties

    override val formInput = MutableStateFlow(FoodItemFormInput.from(submission.item))
    override val recognizedFields = MutableStateFlow<Set<FoodItemFormField>>(emptySet())
    override val isRecognizingNutritionLabel = MutableStateFlow(false)
    override val isNutritionLabelCameraVisible = MutableStateFlow(false)
    override val nutritionLabelCameraHintRes = MutableStateFlow<Int?>(null)
    private val _state = MutableStateFlow<LoadingState<Unit>>(LoadingState.Idle)
    val state: StateFlow<LoadingState<Unit>> = _state
    val alertItem = MutableStateFlow<AlertItem?>(null)
    val isRejectSheetVisible = MutableStateFlow(false)
    val rejectReason = MutableStateFlow("")
    private val _shouldDismiss = MutableStateFlow(false)
    val shouldDismiss: StateFlow<Boolean> = _shouldDismiss
    private val _similarCatalogueItems = MutableStateFlow<List<FoodItemDomain>>(emptyList())
    val similarCatalogueItems: StateFlow<List<FoodItemDomain>> = _similarCatalogueItems
    private val _isSimilarCatalogueItemsSectionAvailable = MutableStateFlow(true)
    val isSimilarCatalogueItemsSectionAvailable: StateFlow<Boolean> = _isSimilarCatalogueItemsSectionAvailable

    val submittedAt: Instant = submission.submittedAt
    val rejectReasonIfAny: String? = submission.rejectReason

    val showsSimilarCatalogueItemsSection: Boolean
        get() = submission.barcode == null

    // MARK: - Functions

    suspend fun onAppear() {
        if (!showsSimilarCatalogueItemsSection) return
        try {
            _similarCatalogueItems.value = searchFoodItems(submission.item.czName)
        } catch (error: CancellationException) {
            throw error
        } catch (error: Exception) {
            Log.warning(error, Constants.LogCategory.MODERATION)
            _isSimilarCatalogueItemsSectionAvailable.value = false
        }
    }

    suspend fun onNutritionLabelCaptured(image: NutritionLabelImage, liveBarcode: String?) {
        recognizeNutritionLabel(image, liveBarcode, recognizeNutritionLabelUseCase)
    }

    fun onNutritionLabelCameraTapped() {
        isNutritionLabelCameraVisible.value = true
    }

    suspend fun onApproveTapped() {
        if (_state.value.isLoading) return
        _state.value = LoadingState.Loading
        val editedItem = formInput.value.asFoodItemDomain(date = submission.item.date)
        try {
            approveSubmission(submission, editedItem)
            onResolved()
            _shouldDismiss.value = true
        } catch (error: CancellationException) {
            throw error
        } catch (_: ApproveSubmissionError.AlreadyResolved) {
            onResolved()
            alertItem.value = AlertItem(titleRes = R.string.moderation_error_alreadyResolved)
            _shouldDismiss.value = true
        } catch (_: ApproveSubmissionError.ChangedSinceReview) {
            onResolved()
            alertItem.value = AlertItem(titleRes = R.string.moderation_error_changedSinceReview)
            _shouldDismiss.value = true
        } catch (error: Exception) {
            Log.error(error, Constants.LogCategory.MODERATION)
            alertItem.value = AlertItem(
                titleRes = when (error) {
                    is CreateFoodItemError.InvalidCode -> R.string.addFood_error_invalidCode
                    is CreateFoodItemError.InvalidName -> R.string.addFood_error_invalidName
                    is CreateFoodItemError.InvalidCalories -> R.string.addFood_error_invalidCalories
                    is CreateFoodItemError.InvalidWeight -> R.string.addFood_error_invalidWeight
                    is CreateFoodItemError.InvalidPortion -> error.error.alertTitleRes
                    is CreateFoodItemError.ItemAlreadyExists -> R.string.moderation_error_alreadyExists
                    else -> R.string.common_error_unknown
                },
            )
        } finally {
            _state.value = LoadingState.loaded
        }
    }

    suspend fun onRejectConfirmed() {
        if (_state.value.isLoading) return
        _state.value = LoadingState.Loading
        try {
            rejectSubmission(submission, rejectReason.value)
            onResolved()
            _shouldDismiss.value = true
        } catch (error: CancellationException) {
            throw error
        } catch (_: RejectSubmissionError.ReasonRequired) {
            alertItem.value = AlertItem(titleRes = R.string.moderation_error_reasonRequired)
        } catch (_: RejectSubmissionError.AlreadyResolved) {
            onResolved()
            alertItem.value = AlertItem(titleRes = R.string.moderation_error_alreadyResolved)
            _shouldDismiss.value = true
        } catch (_: RejectSubmissionError.ChangedSinceReview) {
            onResolved()
            alertItem.value = AlertItem(titleRes = R.string.moderation_error_changedSinceReview)
            _shouldDismiss.value = true
        } catch (error: Exception) {
            Log.error(error, Constants.LogCategory.MODERATION)
            alertItem.value = AlertItem(titleRes = R.string.common_error_unknown)
        } finally {
            _state.value = LoadingState.loaded
        }
    }
}
